components/homme/src/share/cxx/utilities/scream_tridiag.hpp

```diff
@@ -0,0 +1,656 @@
+  assert(X.extent_int(0) == nrow);
+  assert(dl.extent_int(0) == nrow);
+  assert(du.extent_int(0) == nrow);
+  const int tid = impl::get_thread_id_within_team(team);
+  const int nthr = impl::get_team_nthr(team);
+  for (int j = tid; j < nrhs; j += nthr) {
+    for (int i = 1; i < nrow; ++i)
+      X(i,j) -= dl(i) * X(i-1,j);
+    X(nrow-1,j) /= d(nrow-1);
+    for (int i = nrow-1; i > 0; --i)
+      X(i-1,j) = (X(i-1,j) - du(i-1) * X(i,j)) / d(i-1);
+  }
+}
+
+template <typename DT, typename XT>
+KOKKOS_INLINE_FUNCTION
+void thomas_a1x1 (DT* const dl, DT* d, DT* const du, XT* X, const int nrow) {
+  for (int i = 1; i < nrow; ++i) {
+    const auto dli = dl[i] / d[i-1];
+    d[i] -= dli * du[i-1];
+    X[i] -= dli * X[i-1];
+  }
+  X[nrow-1] /= d[nrow-1];
+  for (int i = nrow-1; i > 0; --i)
+    X[i-1] = (X[i-1] - du[i-1] * X[i]) / d[i-1];
+}
+
+template <typename DT, typename XT>
+KOKKOS_INLINE_FUNCTION
+void thomas_a1xm (DT* const dl, DT* d, DT* const du, XT* X,
+                  const int nrow, const int nrhs) {
+  for (int i = 1; i < nrow; ++i) {
+    const auto dli = dl[i] / d[i-1];
+    d[i] -= dli * du[i-1];
+    auto* const xim1 = X + (i-1)*nrhs;
+    auto* const xi = X + i*nrhs;
+    for (int j = 0; j < nrhs; ++j)
+      xi[j] -= dli * xim1[j];
+  }
+  {
+    auto* const xi = X + (nrow-1)*nrhs;
+    for (int j = 0; j < nrhs; ++j)
+      xi[j] /= d[nrow-1];
+  }
+  for (int i = nrow-1; i > 0; --i) {
+    auto* const xim1 = X + (i-1)*nrhs;
+    auto* const xi = X + i*nrhs;
+    for (int j = 0; j < nrhs; ++j)
+      xim1[j] = (xim1[j] - du[i-1] * xi[j]) / d[i-1];
+  }
+}
+
+template <typename DT, typename XT>
+KOKKOS_INLINE_FUNCTION
+void thomas_amxm (DT* const dl, DT* d, DT* const du, XT* X,
+                  const int nrow, const int nrhs) {
+  for (int i = 1; i < nrow; ++i) {
+    const int ios = i*nrhs;
+    const int im1os = (i-1)*nrhs;
+    auto* const dli = dl + ios;
+    auto* const di = d + ios;
+    auto* const dim1 = d + im1os;
+    auto* const duim1 = du + im1os;
+    auto* const xim1 = X + im1os;
+    auto* const xi = X + ios;
+    for (int j = 0; j < nrhs; ++j) {
+      const auto dlij = dli[j] / dim1[j];
+      di[j] -= dlij * duim1[j];
+      xi[j] -= dlij * xim1[j];
+    }
+  }
+  {
+    const int ios = (nrow-1)*nrhs;
+    auto* const di = d + ios;
+    auto* const xi = X + ios;
+    for (int j = 0; j < nrhs; ++j)
+      xi[j] /= di[j];
+  }
+  for (int i = nrow-1; i > 0; --i) {
+    const int ios = i*nrhs;
+    const int im1os = (i-1)*nrhs;
+    auto* const dim1 = d + im1os;
+    auto* const duim1 = du + im1os;
+    auto* const xim1 = X + im1os;
+    auto* const xi = X + ios;
+    for (int j = 0; j < nrhs; ++j)
+      xim1[j] = (xim1[j] - duim1[j] * xi[j]) / dim1[j];
+  }
+}
+
+template <typename TridiagDiag>
+KOKKOS_INLINE_FUNCTION
+void bfb_thomas_factorize (TridiagDiag dl, TridiagDiag d, TridiagDiag du,
+                           typename std::enable_if<TridiagDiag::rank == 1>::type* = 0) {
```

Comment:
Up for debate: I dislike dummy enable_if arguments, aimed at triggering sfinae. I prefer to see them in the template argument, or, better, to call an "impl" function, which takes extra template arguments, which is then specialized on those.

The reason is mostly readability. Extra args need to be parsed before realizing they are just there for sfinae reasons, especially for relatively new developers.

If you don't mind the change from arg to template arg, I can do the mod myself.

Reply:
This is in scream_tridiag.hpp, which is copied in from scream. We can debate this there, but in the Hxx effort, we can't make changes to this file.